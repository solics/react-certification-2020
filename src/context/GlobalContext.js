import React, { createContext, useReducer } from 'react'
import { globalReducer, globalInitialState } from '../reducers'

const GlobalContext = createContext()

function GlobalContextProvider({ children }) {
	const [state, dispatch] = useReducer(globalReducer, globalInitialState)
	return (
		<GlobalContext.Provider value={[state, dispatch]}>{children}</GlobalContext.Provider>
	)
}

export { GlobalContext, GlobalContextProvider }
