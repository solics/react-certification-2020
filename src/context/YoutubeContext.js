import React, { createContext, useState } from 'react'
import { saveArrayOnLocalStorage } from '../utils/localStorage'

const API_KEY = process.env.REACT_APP_YT_API_KEY
const MAX_RESULTS = 20

const YoutubeContext = createContext()

const SEARCH_URL = `https://www.googleapis.com/youtube/v3/search?part=snippet&key=${API_KEY}&type=video&maxResults=${MAX_RESULTS}&`

function YoutubeContextProvider({ children }) {
	const [videos, setVideos] = useState([])
	const [loading, setLoading] = useState(false)
	const [loadingRelated, setLoadingRelated] = useState(false)
	const [errorMsg, setErrorMsg] = useState('')
	const [errorMsgRelated, setErrorMsgRelated] = useState('')
	const [videosRelated, setVideosRelated] = useState([])

	const search = term => {
		setLoading(true)
		fetch(`${SEARCH_URL}q=${term}`)
			.then(response => response.json())
			.then(data => {
				setLoading(false)
				setErrorMsg('')
				if (data?.items?.length) {
					setVideos(data.items)
					saveArrayOnLocalStorage('videos', data.items)
				} else {
					setVideos([])
					setErrorMsg(data.error.message)
				}
			})
			.catch(error => {
				setLoading(false)
				setErrorMsg(error.error.message)
				console.error(error)
			})
	}

	const getRelatedVideos = videoId => {
		setLoadingRelated(true)
		fetch(`${SEARCH_URL}relatedToVideoId=${videoId}`)
			.then(response => response.json())
			.then(data => {
				setLoadingRelated(false)
				setErrorMsgRelated('')
				if (data?.items?.length) {
					setVideosRelated(data.items)
				} else {
					setVideosRelated([])
					setErrorMsgRelated(data.error.message)
				}
			})
			.catch(error => {
				setLoadingRelated(false)
				setErrorMsgRelated(error.error.message)
				console.error(error)
			})
	}

	const defaultContext = {
		videos,
		videosRelated,
		search,
		setVideosContext: newVideos => {
			setVideos(newVideos)
		},
		getRelatedVideos,
		loading,
		loadingRelated,
		errorMsg,
		errorMsgRelated,
	}

	return (
		<YoutubeContext.Provider value={defaultContext}>{children}</YoutubeContext.Provider>
	)
}

export { YoutubeContext, YoutubeContextProvider }
