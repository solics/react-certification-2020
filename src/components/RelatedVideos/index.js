import React, { useEffect, useContext } from 'react'
import styled from 'styled-components/macro'
import { v4 as uuidv4 } from 'uuid'
import { Link } from 'react-router-dom'
import useSearchRelatedVideos from '../../customHooks/useSearchRelatedVideos'
import { GlobalContext } from '../../context/GlobalContext'

const RelatedVideosStyled = styled.div`
	width: 30%;
	a {
		text-decoration: inherit;
	}
	@media only screen and (max-width: 768px) {
		width: 100%;
	}
`
const ThumbnailVideoStyled = styled.div`
	display: grid;
	grid-template-columns: 120px auto;
	margin-bottom: 8px;
	background: ${props => props.theme.headerBackground};
	p {
		margin-top: 8px;
		font-size: 12px;
	}
	span {
		font-size: 10px;
	}
`
const ThumbnailVideoDescriptionStyled = styled.div`
	padding: 0px 8px;
`
export default function RelatedVideos({ videoId, theme }) {
	const [requestRelatedVideos] = useSearchRelatedVideos()
	const [state] = useContext(GlobalContext)

	const {
		youtube: {
			related: { videos },
		},
	} = state

	useEffect(() => {
		requestRelatedVideos(videoId)
	}, [requestRelatedVideos, videoId])

	return videos.length ? (
		<RelatedVideosStyled>
			<h4>Related Videos</h4>
			{videos.map(video => {
				if (!video.snippet) return null
				return (
					<Link key={uuidv4()} to={`/video-detail/${video.id.videoId}`}>
						<ThumbnailVideoStyled theme={theme}>
							<img
								src={video.snippet?.thumbnails.default.url}
								alt={video.snippet?.title || 'alt-description'}
							/>
							<ThumbnailVideoDescriptionStyled>
								<p>{video.snippet?.title}</p>
								<span>{video.snippet?.channelTitle}</span>
							</ThumbnailVideoDescriptionStyled>
						</ThumbnailVideoStyled>
					</Link>
				)
			})}
		</RelatedVideosStyled>
	) : (
		<div>loading...</div>
	)
}
