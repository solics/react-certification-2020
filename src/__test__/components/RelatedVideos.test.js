import { render } from '@testing-library/react'
import React from 'react'
import { GlobalContextProvider } from '../../context/GlobalContext'
import RelatedVideos from '../../components/RelatedVideos'
import THEMES from '../../themes'

describe('VideosList testing', () => {
	it('VideosList Snapshot', () => {
		const { container } = render(
			<GlobalContextProvider>
				<RelatedVideos videoId="1123asdsa123" theme={THEMES.light} />
			</GlobalContextProvider>
		)
		expect(container).toMatchSnapshot()
	})
})
