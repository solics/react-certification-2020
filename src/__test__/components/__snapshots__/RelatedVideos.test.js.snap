// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideosList testing VideosList Snapshot 1`] = `
<div>
  <div>
    loading...
  </div>
</div>
`;
